File: LeetCodePractice/src/Math/SmallestIntegerDivisibleByK.java
package Math;
/*
Given a positive integer K, you need to find the length of the smallest positive integer N such that N is divisible by K, and N only contains the digit 1.

Return the length of N. If there is no such N, return -1.

Note: N may not fit in a 64-bit signed integer.



Example 1:

Input: K = 1
Output: 1
Explanation: The smallest answer is N = 1, which has length 1.
Example 2:

Input: K = 2
Output: -1
Explanation: There is no such positive integer N divisible by 2.
Example 3:

Input: K = 3
Output: 3
Explanation: The smallest answer is N = 111, which has length 3.


Constraints:

1 <= K <= 10^5

analysis:
Furthermore, we can improve this algorithm with Pigeonhole Principle.
Recall that the number of possible values of remainder (ranging from 0 to K-1) is limited, and in fact, the number is K.
As a result, if the while-loop continues more than K times, and haven't stopped,
then we can conclude that remainder repeats -- you can not have more than K different remainder.
 */
public class SmallestIntegerDivisibleByK {
    public int smallestRepunitDivByK(int K) {
        int remainder = 0;
        for(int i = 1; i <= K; i++){
            remainder = (remainder * 10 + 1) % K;
            if(remainder == 0){
                return i;
            }
        }
        return -1;
    }
}

File: LeetCodePython/SmallestIntegerDivisibleByK.py
"""
Given a positive integer k, you need to find the length of the smallest positive integer n such that n is divisible by k, and n only contains the digit 1.

Return the length of n. If there is no such n, return -1.

Note: n may not fit in a 64-bit signed integer.

Example 1:
Input: k = 1
Output: 1
Explanation: The smallest answer is n = 1, which has length 1.

Example 2:
Input: k = 2
Output: -1
Explanation: There is no such positive integer n divisible by 2.

Example 3:
Input: k = 3
Output: 3
Explanation: The smallest answer is n = 111, which has length 3.

Constraints:
1 <= k <= 10^5

hints:
1 11111 = 1111 * 10 + 1 We only need to store remainders modulo K.
2 If we never get a remainder of 0, why would that happen, and how would we know that?

Analysis:
Furthermore, we can improve this algorithm with Pigeonhole Principle.
Recall that the number of possible values of remainder (ranging from 0 to K-1) is limited, and in fact, the number is K.
As a result, if the while-loop continues more than K times, and haven't stopped,
then we can conclude that remainder repeats -- you can not have more than K different remainder.

TC: O(K)
"""

class SmallestIntegerDivisibleByK:
    def smallestRepunitDivByK(self, k: int) -> int:
        res = cur = 1
        for _ in range(k):
            if cur % k == 0:
                return res
            res += 1
            cur = cur * 10 + 1
        return -1